Add countdown_timer.py

JerryWorkman opened the pull request:
**Description:**
Combines one or more binary_sensors (typically PIR sensors) 
and a switch to provide a switch that automatically
turns off after a specified delay time. 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1697

**Example entry for `configuration.yaml` (if applicable):**
```yaml
# Example configuration.yaml entry
sensor countdown_timer:
  platform: countdown_timer
  timers:
    lr_light:
      sensors: binary_sensor.lr_pir1
      switch:  switch.lr_light_switch
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

Commit:
Add countdown_timer.py

andrey-git commented on homeassistant/components/sensor/countdown_timer.py:
```diff
@@ -0,0 +1,172 @@
+})
+
+
+def _debug(msg):
```
I think logger already has the domain from __name__.

pvizeli commented:
Platform or component without device should be async or it will run into executor deadlock. Look at async guide on our website.

pvizeli reviewed (changes requested):
Dont use thread timer. Use our time change stuff in helpers/events

Ornias1993 commented:
Was viewing this, but cant keep myself from commenting @pvizeli 
Next time you point someone to your documentation you should be drowned in public. It is so damned scarce, you should be at the very least ashamed.

If you follow home-assistant related custom component development, you should have seen that complaint 10000 times already.

Ornias1993 commented:
It wont be that hard to just post a link on the website for example on a clear location to:
https://dev-docs.home-assistant.io/en/master/

JerryWorkman commented:
The documentation pull request hasn't been merged yet. Until then you can
find the docs here:
https://github.com/JerryWorkman/home-assistant.github.io/blob/next/source/_components/sensor.countdown_timer.markdown

Still learning the ropes...

On Mon, Jan 2, 2017 at 5:02 AM, Kjeld Schouten <notifications@github.com>
wrote:

> Was viewing this, but cant keep myself from commenting @pvizeli
> <https://github.com/pvizeli>
> Next time you point someone to your documentation you should be drowned in
> public. It is so damned scarce, you should be at the very least ashamed.
>
> If you follow home-assistant related custom component development, you
> should have seen that complaint 10000 times already.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/5137#issuecomment-269952227>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQMPaEdVMoELnK4c9ifNZWNOMMEhLks5rOMs4gaJpZM4LY11V>
> .
>

fabaff commented:
@Ornias1993 We are not talking about custom component development here, it's about to include stuff into Home Assistant and @pvizeli has every right to be short and precise. Especially as it's the third time a topic is coming up...

> If you follow home-assistant related custom component development, you should have seen that complaint 10000 times already.

If there are 10000 complains then I'm wondering where the the open issues or the PRs to fix it are...

Ornias1993 commented:
> @Ornias1993 We are not talking about custom component development here, it's about to include stuff 

Ohh, its not a custom component, whats it then... a rainbow farting unicorn?! >.<

"Then I'm wondering where the the open issues or the PRs to fix it are..."
never taken a look into the component documentation for HA... thats clear.
Maybe the 10000 is a bit on the big side, most people give up after figuringout un-documented function number 1000.

Sorry, but your documentation is NOT hobby-dev friendly. Yes professional python devs will be perfectly able to use it, but if you need that degree of competence due to bad documentation, there is something wrong with your documentation.

Danielhiversen commented:
@Ornias1993 
I think we should keep this discussion out of this pull request. 

But it would be great if you help update and improve the documentation. In the upper right corner of any documentation page, you will find "Edit this page on GitHub". Follow the link and submit your suggested change. Demonstrated here: https://youtu.be/dRfk9JAlPJk?t=1h16m17s

JerryWorkman committed:
Update countdown_timer.py

Replace Python Timer with HA track_point_in_time.
This component does not do any I/O so asyncio is not required.

JerryWorkman reviewed (commented):
Timer was replaced with track_point_in_time

balloob commented:
@Ornias1993 your tone in your comments is highly inappropriate. Telling someone to "drowned in public" is not allowed within our community. I have blocked you from further participation from our GitHub organization.

balloob commented:
@JerryWorkman you are trying to write an automation component but have it be disguised as a sensor platform. With automation I don't mean the actual automation component, but more an automation template.

You should model your automation after for example the [`device_sun_light_trigger`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/device_sun_light_trigger.py) automation component. All automation components have to be async and accompanied with tests.

So for this PR to get accepted, you should:

 - Migrate your sensor platform to a component
 - Make it async
 - Write tests

JerryWorkman commented:
Thanks.

On Tue, Jan 17, 2017 at 2:27 AM, Paulus Schoutsen <notifications@github.com>
wrote:

> @JerryWorkman <https://github.com/JerryWorkman> you are trying to write
> an automation component but have it be disguised as a sensor platform. With
> automation I don't mean the actual automation component, but more an
> automation template.
>
> You should model your automation after for example the
> device_sun_light_trigger
> <https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/device_sun_light_trigger.py>
> automation component. All automation components have to be async and
> accompanied with tests.
>
> So for this PR to get accepted, you should:
>
>    - Migrate your sensor platform to a component
>    - Make it async
>    - Write tests
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/5137#issuecomment-273040900>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQDF72X3zjLcFhOe-SxA5zLCCZSdTks5rTG1YgaJpZM4LY11V>
> .
>

JerryWorkman commented:
Where can I example tests?

On Tue, Jan 17, 2017 at 12:18 PM, Jerry Workman <jerry.workman@gmail.com>
wrote:

> Thanks.
>
> On Tue, Jan 17, 2017 at 2:27 AM, Paulus Schoutsen <
> notifications@github.com> wrote:
>
>> @JerryWorkman <https://github.com/JerryWorkman> you are trying to write
>> an automation component but have it be disguised as a sensor platform. With
>> automation I don't mean the actual automation component, but more an
>> automation template.
>>
>> You should model your automation after for example the
>> device_sun_light_trigger
>> <https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/device_sun_light_trigger.py>
>> automation component. All automation components have to be async and
>> accompanied with tests.
>>
>> So for this PR to get accepted, you should:
>>
>>    - Migrate your sensor platform to a component
>>    - Make it async
>>    - Write tests
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/home-assistant/home-assistant/pull/5137#issuecomment-273040900>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABsbQDF72X3zjLcFhOe-SxA5zLCCZSdTks5rTG1YgaJpZM4LY11V>
>> .
>>
>
>

balloob commented:
Example tests can be found here: https://github.com/home-assistant/home-assistant/blob/dev/tests/components/test_device_sun_light_trigger.py

JerryWorkman commented:
Thanks Again.

On Wed, Jan 18, 2017 at 12:04 AM, Paulus Schoutsen <notifications@github.com
> wrote:

> Example tests can be found here: https://github.com/home-
> assistant/home-assistant/blob/dev/tests/components/test_
> device_sun_light_trigger.py
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/5137#issuecomment-273383852>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQAQ7iDubDHe77F-keeEQSYO7jYqTks5rTZ1qgaJpZM4LY11V>
> .
>

homeassistant commented:
Hello @JerryWorkman,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`832e1217f29b4b646e5b1d494497519d3fbd1c97`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;5137&#x2F;commits&#x2F;832e1217f29b4b646e5b1d494497519d3fbd1c97) - No email found attached to the commit.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

andrey-git commented:
@JerryWorkman any updates on this PR?

JerryWorkman commented:
I thought that I dropped it. If not kill the PR.

I will post it as a "Share your Project" soon and perhaps try again layer
as a component. I just don't have the time to write the testing code and
change the project from a sensor to a bare component as requested.

It works great now but doesn't yet comply with new component specs.

Jerry

On Tue, Feb 7, 2017 at 2:54 PM, Andrey <notifications@github.com> wrote:

> @JerryWorkman <https://github.com/JerryWorkman> any updates on this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/5137#issuecomment-278120552>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQPm9R3l_unfb14T_lXRmr1Z2ckJBks5raMvhgaJpZM4LY11V>
> .
>

andrey-git commented:
Thanks, Jerry!
